metadata-ingestion/src/datahub/ingestion/source/gcs/gcs_source.py

```diff
@@ -88,7 +89,10 @@ def __init__(self, config: GCSSourceConfig, ctx: PipelineContext):
         super().__init__(config, ctx)
         self.config = config
         self.report = GCSSourceReport()
-        self.s3_source = self.create_equivalent_s3_source(ctx)
+        self.platform: str = "gcs"
+        s3_ctx = copy.deepcopy(ctx)
+        s3_ctx.pipeline_name = None
```

Comment:
You may move these two lines inside `create_equivalent_s3_source` 

Also, how is `s3_ctx.pipeline_name = None` is required?
Note `pipeline_name` is required when `stateful_ingestion` is enabled.

Reply:
Without this, running stateful ingestion leads to `IndexError: Checkpointing provider DatahubIngestionCheckpointingProvider already registered.`. This is because `StatefulIngestionSourceBase` is initialized twice, first for `GCSSource` and then in `create_equivalent_s3_source` with `S3Source`. This can also be verified by the change I did in the unit test.
I think, stateful ingestion was always meant to be switched off for the equivalent S3 source object, because in `create_equivalent_s3_config`, `DataLakeSourceConfig` is initialized without `stateful_ingestion`, which implicitly means `stateful_ingestion=None`. Since `pipeline_name` is set, the logic in `StateProviderWrapper` leads to stateful ingestion being switched on, which leads to the error.

I put the lines in the function, as suggested.